app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,419 @@
+            startTimestampUs =
+                    System.nanoTime() / RecorderConstant.ENCODER_PRESENTATION_TIME_DEFAULT_DIVIDER;
+            startTimestampInitialized = true;
+        }
+        return (System.nanoTime() / RecorderConstant.ENCODER_PRESENTATION_TIME_DEFAULT_DIVIDER
+                        - startTimestampUs);
+    }
+
+    private int calcBitRate(int width, int height) {
+        final int bitrate = (int) (RecorderConstant.BITRATE_MULTIPLIER *
+                RecorderConstant.VIDEO_CODEC_FRAME_RATE * width * height);
+        Log.i(TAG, String.format("bitrate=%5.2f[Mbps]", bitrate / 1024f / 1024f));
+        return bitrate;
+    }
+
+    private void startMuxerIfSetUp() {
+        if (audioTrackIndex >= 0 && videoTrackIndex >= 0) {
+            muxer.start();
+            muxerStarted = true;
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private boolean writeAudioBufferToFile() {
+        int encoderStatus;
+
+        encoderStatus = audioEncoder.dequeueOutputBuffer(this.bufferInfo, 0);
+        if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
+            if (audioTrackIndex > 0) {
+                Log.e(TAG, "audioTrackIndex less than zero");
+                return false;
+            }
+            audioTrackIndex = muxer.addTrack(audioEncoder.getOutputFormat());
+            startMuxerIfSetUp();
+        } else if (encoderStatus < 0 && encoderStatus != MediaCodec.INFO_TRY_AGAIN_LATER) {
+            Log.w(TAG, "unexpected result from audio encoder.dequeueOutputBuffer: "
+                    + encoderStatus);
+        } else if (encoderStatus >= 0) {
+            ByteBuffer encodedData = audioEncoder.getOutputBuffer(encoderStatus);
+            if (encodedData == null) {
+                Log.e(TAG, "encodedData null");
+                return false;
+            }
+
+            if (this.bufferInfo.presentationTimeUs > this.lastAudioTimestampUs
+                    && muxerStarted && this.bufferInfo.size != 0 &&
+                    (this.bufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) == 0) {
+                this.lastAudioTimestampUs = this.bufferInfo.presentationTimeUs;
+                muxer.writeSampleData(audioTrackIndex, encodedData, this.bufferInfo);
+            }
+
+            audioEncoder.releaseOutputBuffer(encoderStatus, false);
+
+            if ((this.bufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
+                Log.v(TAG, "buffer eos");
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private boolean writeVideoBufferToFile() {
+        int encoderStatus;
+
+        encoderStatus = videoEncoder.dequeueOutputBuffer(bufferInfo,
+                RecorderConstant.MEDIA_QUEUE_BUFFERING_DEFAULT_TIMEOUT);
+        if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
+            if (videoTrackIndex > 0) {
```

Comment:
this comparison does not match to the log message below

Reply:
done